crypto/ec/ec_evp_lib.c

```diff
@@ -0,0 +1,417 @@
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_TYPE, kdf, NULL);
+
+    /* Check the ctx actually understands this parameter */
+    if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_settable_params(ctx),
+                OSSL_EXCHANGE_PARAM_KDF_TYPE) == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
+
+    /* TODO(romen) : magic numbers or strings toward provider? */
+    *p++ = OSSL_PARAM_construct_int(OSSL_EXCHANGE_PARAM_KDF_TYPE,
+                                    &kdf);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (!EVP_PKEY_CTX_set_params(ctx, params))
+        return 0;
+
+    return 1;
+}
+
+/* TODO(3.0): should constify? */
+int EVP_PKEY_CTX_get_ecdh_kdf_type(EVP_PKEY_CTX *ctx)
+{
+    int ret;
+    OSSL_PARAM params[2], *p = params;
+
+    if ( 1 != (ret = evp_pkey_ctx_getset_ecdh_param_checks(ctx)))
+        return ret;
+
+    /* TODO(3.0): Remove this eventually when no more legacy */
+    if (ctx->op.kex.exchprovctx == NULL)
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_TYPE, -2, NULL);
+
+    /* Check the ctx actually understands this parameter */
+    if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_gettable_params(ctx),
+                OSSL_EXCHANGE_PARAM_KDF_TYPE) == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
+
+    *p++ = OSSL_PARAM_construct_int(OSSL_EXCHANGE_PARAM_KDF_TYPE,
+                                    &ret);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (!EVP_PKEY_CTX_get_params(ctx, params))
+        return -1;
+
+    /* TODO(romen) : magic numbers or strings toward provider? */
+    if (ret != EVP_PKEY_ECDH_KDF_NONE && ret != EVP_PKEY_ECDH_KDF_X9_63)
+        return -2;
+
+    return ret;
+}
+
+int EVP_PKEY_CTX_set_ecdh_kdf_md(EVP_PKEY_CTX *ctx, const EVP_MD *md)
+{
+    int ret;
+    OSSL_PARAM params[2], *p = params;
+    const char *md_name = NULL;
+
+    if ( 1 != (ret = evp_pkey_ctx_getset_ecdh_param_checks(ctx)))
+        return ret;
+
+    /* TODO(3.0): Remove this eventually when no more legacy */
+    if (ctx->op.kex.exchprovctx == NULL)
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_MD, 0, (void *)(md));
+
+    /* Check the ctx actually understands this parameter */
+    if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_settable_params(ctx),
+                OSSL_EXCHANGE_PARAM_KDF_MD) == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
```

Comment:
I am doing this basically everywhere. Is there a more clever way of calling `EVP_PKEY_CTX_set_params` in a way that fails if all the parameters are not supported by the provider?

It would reduce a lot of code duplication if we had a dedicated `EVP_PKEY_CTX_set_params_strict` (and the getter counterpart).

What are your thoughts about it?

Reply:
Do we actually need to check that the  provider understands these parameters? One possibility is to just let the provider do what it wants with these values (which maybe to ignore them). If we *must* know that the provider has understood it then I'd be ok with creating a dedicate helper function to do this.

Reply:
I would say that existing applications in 1.1.1 expect the function call to fail if the requested thing is not handled by the backend.
The current default of ignoring unhandled things is potentially a breaking change (and we have very little negative tests for this kind of things to measure how much of a breakage they could cause).

Reply:
If you really need to do it use gettable_params to query which params are supported.

Reply:
Why gettable_params? In this case we are setting so I am checking against settable_params

Reply:
Yep I am wrong.. Do what matt said :)

Reply:
I went ahead and added local helpers to do this.

> Do we actually need to check that the provider understands these parameters? One possibility is to just let the provider do what it wants with these values (which maybe to ignore them). If we must know that the provider has understood it then I'd be ok with creating a dedicate helper function to do this.

@mattcaswell 
For cofactor ECDH I see it as a requirement for security, for all the other `EVP_PKEY_CTX_set_foo()` I see it as a backward compatibility necessity.

Existing apps upon calling `EVP_PKEY_CTX_set_foo()`/`EVP_PKEY_CTX_ctrl()` rely on the callee returning an error if the request was invalid or not handled. If we allow providers to silently ignore these requests we can't conform to the existing behavior.

But given you and @slontis have not encountered this need in other ports I am starting to think that maybe I am missing something fundamental that explains why you don't think this is needed here and elsewhere.


For now I added local helpers to do this systematically and reduce repeated code. In case mine is deemed the right approach, we might want to discuss if these helpers are needed elsewhere internally to promote them to internal APIs.

Reply:
The local helpers have now been promoted to internal functions, and moved in `crypto/evp/pmeth_lib.c`